tests/python/relay/test_ir_parser.py

```diff
@@ -632,6 +641,236 @@ def test_tuple_type():
+        rest_var = relay.Var("rest")
+        cons_case = relay.Let(
+            _,
+            UNIT,
+            relay.add(relay.const(1), relay.Call(length_var, [rest_var])))
+        body = relay.Match(input_var,
+            [relay.Clause(
+                relay.PatternConstructor(
+                    cons_constructor,
+                    [relay.PatternWildcard(), relay.PatternVar(rest_var)]),
+                cons_case),
+            relay.Clause(
+                relay.PatternConstructor(nil_constructor, []),
+                relay.const(0))],
+            complete=is_complete
+        )
+        length_func = relay.Function(
+            [input_var],
+            body,
+            int32,
+            [typ_var]
+        )
+        mod[length_var] = length_func
+
+        assert parses_as(
+            """
+            %s
+
+            def @length[A](%%xs: List[A]) -> int32 {
+              %s (%%xs) {
+                Cons(_, %%rest) => {
+                  ();;
+                  1 + @length(%%rest)
+                },
+                Nil => 0,
+              }
+            }
+            """ % (LIST_DEFN, match_keyword),
+            mod
+        )
+
+
+def test_adt_cons_expr():
+    mod = relay.Module()
+
+    list_var = relay.GlobalTypeVar("List")
+    typ_var = relay.TypeVar("A")
+    cons_constructor = relay.Constructor(
+        "Cons", [typ_var, list_var(typ_var)], list_var)
+    nil_constructor = relay.Constructor("Nil", [], list_var)
+    list_def = relay.TypeData(
+        list_var,
+        [typ_var],
+        [cons_constructor, nil_constructor])
+    mod[list_var] = list_def
+
+    make_singleton_var = relay.GlobalVar("make_singleton")
+    input_var = relay.Var("x", int32)
+    make_singleton_func = relay.Function(
+        [input_var],
+        cons_constructor(input_var, nil_constructor()),
+        list_var(int32)
+    )
+    mod[make_singleton_var] = make_singleton_func
+
+    assert parses_as(
+        """
+        %s
+
+        def @make_singleton(%%x: int32) -> List[int32] {
+          Cons(%%x, Nil())
+        }
+        """ % LIST_DEFN,
+        mod
+    )
+
+@raises_parse_error
+def test_duplicate_adt_defn():
+    parse_text(
+        """
+        %s
+
+        type List[A] {
+          Cons(A, List[A]),
+          Nil,
+        }
+        """ % LIST_DEFN
+    )
+
+
+@raises_parse_error
+def test_duplicate_adt_cons():
+    parse_text(
+        """
+        type Ayy { Lmao }
+        type Haha { Lmao }
+        """
+    )
+
+
+@raises_parse_error
+def test_duplicate_adt_cons_defn():
+    parse_text(
+        """
+        type Ayy { Lmao }
+        type Lmao { Ayy }
+        """
+    )
+
+
+@raises_parse_error
+def test_duplicate_global_var():
+    parse_text(
+        """
+        def @id[A](%x: A) -> A { x }
+        def @id[A](%x: A) -> A { x }
+        """
+    )
+
+
+def test_extern_adt_defn():
+    # TODO(weberlo): update this test once extern is implemented
```

Comment:
Maybe we shouldn't include the syntax while we don't have the feature implemented (e.g., leave it for a different PR)

Reply:
@jroesch is planning on updating it very soon for the VM.  He just wanted me to add all of the necessary boilerplate.

Reply:
okay

Reply:
I need this change ASAP, I just want the parsing changes to be shipped in a unit to keep things clean.